app/Http/Controllers/ScoresController.php

```diff
@@ -57,21 +57,20 @@ public function download($rulesetOrSoloId, $id = null)
                 return ujs_redirect(route('scores.show', ['rulesetOrScore' => $rulesetOrSoloId]));
             }
             $soloScore = SoloScore::where('has_replay', true)->findOrFail($rulesetOrSoloId);
-
-            $score = $soloScore->legacyScore() ?? $soloScore;
         } else {
             if ($shouldRedirect) {
                 return ujs_redirect(route('scores.show', ['rulesetOrScore' => $rulesetOrSoloId, 'score' => $id]));
             }
             // don't limit downloading replays of restricted users for review purpose
-            $score = ScoreBest::getClass($rulesetOrSoloId)
-                ::where('score_id', $id)
-                ->where('replay', true)
-                ->firstOrFail();
-
-            $soloScore = SoloScore::firstWhere(['legacy_score_id' => $score->getKey(), 'ruleset_id' => $score->ruleset_id]);
+            $soloScore = SoloScore::where([
+                'has_replay' => true,
+                'legacy_score_id' => $id,
+                'ruleset_id' => Beatmap::MODES[$rulesetOrSoloId],
+            ])->firstOrFail();
         }
 
+        $score = $soloScore->legacyScore() ?? $soloScore;
```

Comment:
Just noticed the lock key later on is not including the ruleset for legacy scores